server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1180,151 +916,315 @@ public void snapshot(final IndexCommit snapshotIndexCommit) {
+                    indexShardSnapshotsFormat.read(shardContainer, Integer.toString(latest));
+                return new Tuple<>(shardSnapshots, latest);
+            } catch (IOException e) {
+                final String file = SNAPSHOT_INDEX_PREFIX + latest;
+                logger.warn(() -> new ParameterizedMessage("failed to read index file [{}]", file), e);
+            }
+        } else if (blobKeys.isEmpty() == false) {
+            logger.warn("Could not find a readable index-N file in a non-empty shard snapshot directory [{}]", shardContainer.path());
+        }
 
-            private void checkAborted() {
-                if (snapshotStatus.isAborted()) {
-                    logger.debug("[{}] [{}] Aborted on the file [{}], exiting", shardId, snapshotId, fileName);
-                    throw new IndexShardSnapshotFailedException(shardId, "Aborted");
+        // We couldn't load the index file - falling back to loading individual snapshots
+        List<SnapshotFiles> snapshots = new ArrayList<>();
+        for (String name : blobKeys) {
+            try {
+                BlobStoreIndexShardSnapshot snapshot = null;
+                if (name.startsWith(SNAPSHOT_PREFIX)) {
+                    snapshot = indexShardSnapshotFormat.readBlob(shardContainer, name);
                 }
+                if (snapshot != null) {
+                    snapshots.add(new SnapshotFiles(snapshot.snapshot(), snapshot.indexFiles()));
+                }
+            } catch (IOException e) {
+                logger.warn(() -> new ParameterizedMessage("failed to read commit point [{}]", name), e);
             }
         }
+        return new Tuple<>(new BlobStoreIndexShardSnapshots(snapshots), latest);
     }
 
-    private static final class PartSliceStream extends SlicedInputStream {
+    /**
+     * Snapshot individual file
+     * @param fileInfo file to be snapshotted
+     */
+    private void snapshotFile(BlobStoreIndexShardSnapshot.FileInfo fileInfo, IndexId indexId, ShardId snapshotShardId,
+                              SnapshotId snapshotId, IndexShardSnapshotStatus snapshotStatus, Store store) throws IOException {
+        final BlobContainer shardContainer = shardContainer(indexId, snapshotShardId);
+        final String file = fileInfo.physicalName();
+        try (IndexInput indexInput = store.openVerifyingInput(file, IOContext.READONCE, fileInfo.metadata())) {
+            for (int i = 0; i < fileInfo.numberOfParts(); i++) {
+                final long partBytes = fileInfo.partBytes(i);
+
+                InputStream inputStream = new InputStreamIndexInput(indexInput, partBytes);
+                if (snapshotRateLimiter != null) {
+                    inputStream = new RateLimitingInputStream(inputStream, snapshotRateLimiter,
+                        snapshotRateLimitingTimeInNanos::inc);
+                }
+                inputStream = new FilterInputStream(inputStream) {
```

Comment:
By inlining the AbortableInputStream class we lost the "abortable" name that made the role of the input stream obvious. Can you please add a short comment?

Reply:
Added :)